docs/03-Contributing/08-API-Versioning.md

```diff
@@ -0,0 +1,104 @@
+# API Versioning
+
+## Why do we need to version the API?
+
+The API is a product, just like fleetctl, and the web UI. It has its users, mostly fleetctl and the web UI but there are
+also third party developers working with it.
+
+Evolving a product inherently needs versioning. Most products create a new version with any addition to the product, but 
+the API will work differently in that regard as new additions to the API won't increase the version.
+
+New versions will be used for breaking changes and deprecating APIs. Only when a breaking change is introduced we will 
+release a new version of the API.
+
+## What kind of versioning will we use for the API?
+
+The format for the API version we've chosen is that of a date with the following format:
+
+```
+<year>-<month>
+```
+
+The date is chosen based on the month the breaking change was introduced.
+
+## Why is v1 still available at the time of this writing?
+
+`v1` is the first version of the API. It existed before this text, and so it doesn't follow the versioning schema 
+explained here. We still need to support it for a few months (see below on deprecation). So it'll be treated as an 
+exception in the logic in the Go code while it exists.
+
+## Why not semantic versioning?
+
+Semantic versioning is great and we are using it in Fleet itself. However, it doesn't necessarily work for APIs since we 
+are not going to be releasing a new version with every addition, just with breaking changes. So it doesn't align with our 
+needs in the API level.
+
+## How are API releases aligned with regular Fleet releases?
+
+New versions are deployed when Fleet is released, given the nature of the product. However, not all new versions of 
+Fleet will have a new release for the API.
+
+## How long do I have until you remove a deprecated API?
+
+3 months after the new release has been available.
```

Comment:
Here is a really good read from Swagger on the importance of taking care when deprecating APIs: https://swagger.io/blog/api-strategy/best-practices-for-deprecating-apis/

The relevant points:

>Remember that if deprecating APIs is not done correctly, it can damage a company’s brand reputation and trust.> 

>Provide a long enough sunset period. This is the period after the initial announcement to deprecate APIs, that gives API consumers time to reconfigure their apps. This could be in the form of code updates to fit the new version of the API into their software architecture, or a program to accommodate alternate solutions in case the API has been disabled. A good sunset period is important as it gives API-dependent companies time to work on changing their business and technological strategy. Depending on the deprecated API’s reach, user base and service offering, this period could be anywhere between 3 - 8 months.>

I think we should have a variable timeframe on API version deprecation based on usage and complexity of the breaking change. 3 months seems sufficient for small changes, but larger changes on more heavily used releases will likely require more time to make sure we're not pulling the rug on anyone, especially as the API consumer base grows.

I'd also like to work closely with @ksatter any time we deprecate an API version to ensure we're providing a good developer experience.

Reply:
> I think we should have a variable timeframe on API version deprecation based on usage and complexity of the breaking change. 3 months seems sufficient for small changes, but larger changes on more heavily used releases will likely require more time to make sure we're not pulling the rug on anyone, especially as the API consumer base grows.

Well, we wouldn't be pulling the rug on anybody as they can choose when to upgrade.

If we do variable deprecation timeline, it's less predictive and can lend itself to dragging a version for too long. Would 6 months be a more reasonable timeframe?

Reply:
Good point, no APIs will ever disappear on them since they control the deployment. 

6 months sounds like a good timeline if we want to have a consistent timeline for all releases.